// @pega/custom-dx-components/tests/assets/components/createAll/Launchpad/Pega_DXIL_MyCaseWidget/mock.ts
// @ts-nocheck
const historyData = {
  data: {
    fetchDateTime: '2023-01-18T12:04:56.517Z',
    pxObjClass: 'Pega-API-DataExploration-Data',
    resultCount: 31,
    data: [
      {
        CreateDateTime: '2023-01-13T21:46:33.339Z',
        pxObjClass: 'History-DXIL-DigV2-Work',
        PerformerID: 'User DigV2',
        pxInsName: 'DXIL-DIGV2-WORK C-7430!20230113T214633.339 GMT',
        pxLongitude: null,
        pzInsKey: 'HISTORY-DXIL-DIGV2-WORK DXIL-DIGV2-WORK C-7430!20230113T214633.339 GMT',
        pxHistoryForReference: 'DXIL-DIGV2-WORK C-7430',
        MessageText: "Assigned to User DigV2 to 'complete task' ",
        pyMemo: null,
        pxLatitude: null
      },
      {
        CreateDateTime: '2023-01-13T21:46:33.332Z',
        pxObjClass: 'History-DXIL-DigV2-Work',
        PerformerID: 'User DigV2',
        pxInsName: 'DXIL-DIGV2-WORK C-7430!20230113T214633.332 GMT',
        pzInsKey: 'HISTORY-DXIL-DIGV2-WORK DXIL-DIGV2-WORK C-7430!20230113T214633.332 GMT',
        pxLongitude: null,
        MessageText: "Assignment to 'complete task' completed by performing a 'EmbeddedData'.",
        pxHistoryForReference: 'DXIL-DIGV2-WORK C-7430',
        pyMemo: null,
        pxLatitude: null
      },
      {
        CreateDateTime: '2023-01-13T21:46:32.005Z',
        pxObjClass: 'History-DXIL-DigV2-Work',
        PerformerID: 'User DigV2',
        pxInsName: 'DXIL-DIGV2-WORK C-7430!20230113T214632.005 GMT',
        pzInsKey: 'HISTORY-DXIL-DIGV2-WORK DXIL-DIGV2-WORK C-7430!20230113T214632.005 GMT',
        pxLongitude: null,
        pxHistoryForReference: 'DXIL-DIGV2-WORK C-7430',
        MessageText: "Assigned to User DigV2 to 'complete task' ",
        pyMemo: null,
        pxLatitude: null
      },
      {
        CreateDateTime: '2023-01-13T21:46:31.998Z',
        pxObjClass: 'History-DXIL-DigV2-Work',
        PerformerID: 'User DigV2',
        pxInsName: 'DXIL-DIGV2-WORK C-7430!20230113T214631.998 GMT',
        pzInsKey: 'HISTORY-DXIL-DIGV2-WORK DXIL-DIGV2-WORK C-7430!20230113T214631.998 GMT',
        pxLongitude: null,
        pxHistoryForReference: 'DXIL-DIGV2-WORK C-7430',
        MessageText: "Assignment to 'complete task' removed via ticket.",
        pxLatitude: null,
        pyMemo: null
      },
      {
        CreateDateTime: '2023-01-13T21:46:31.175Z',
        pxObjClass: 'History-DXIL-DigV2-Work',
        PerformerID: 'User DigV2',
        pxInsName: 'DXIL-DIGV2-WORK C-7430!20230113T214631.175 GMT',
        pxLongitude: null,
        pzInsKey: 'HISTORY-DXIL-DIGV2-WORK DXIL-DIGV2-WORK C-7430!20230113T214631.175 GMT',
        MessageText: "Assigned to User DigV2 to 'complete task' ",
        pxHistoryForReference: 'DXIL-DIGV2-WORK C-7430',
        pxLatitude: null,
        pyMemo: null
      },
      {
        CreateDateTime: '2023-01-13T21:46:31.168Z',
        pxObjClass: 'History-DXIL-DigV2-Work',
        PerformerID: 'User DigV2',
        pxInsName: 'DXIL-DIGV2-WORK C-7430!20230113T214631.168 GMT',
        pzInsKey: 'HISTORY-DXIL-DIGV2-WORK DXIL-DIGV2-WORK C-7430!20230113T214631.168 GMT',
        pxLongitude: null,
        pxHistoryForReference: 'DXIL-DIGV2-WORK C-7430',
        MessageText: "Assignment to 'complete task' completed by performing a 'EmbeddedData'.",
        pyMemo: null,
        pxLatitude: null
      },
      {
        pxObjClass: 'History-DXIL-DigV2-Work',
        CreateDateTime: '2023-01-13T21:46:29.893Z',
        PerformerID: 'User DigV2',
        pxInsName: 'DXIL-DIGV2-WORK C-7430!20230113T214629.893 GMT',
        pzInsKey: 'HISTORY-DXIL-DIGV2-WORK DXIL-DIGV2-WORK C-7430!20230113T214629.893 GMT',
        pxLongitude: null,
        pxHistoryForReference: 'DXIL-DIGV2-WORK C-7430',
        MessageText: "Assigned to User DigV2 to 'complete task' ",
        pyMemo: null,
        pxLatitude: null
      },
      {
        pxObjClass: 'History-DXIL-DigV2-Work',
        CreateDateTime: '2023-01-13T21:46:29.887Z',
        PerformerID: 'User DigV2',
        pxInsName: 'DXIL-DIGV2-WORK C-7430!20230113T214629.887 GMT',
        pxLongitude: null,
        pzInsKey: 'HISTORY-DXIL-DIGV2-WORK DXIL-DIGV2-WORK C-7430!20230113T214629.887 GMT',
        pxHistoryForReference: 'DXIL-DIGV2-WORK C-7430',
        MessageText: "Assignment to 'complete task' removed via ticket.",
        pxLatitude: null,
        pyMemo: null
      },
      {
        CreateDateTime: '2023-01-13T21:46:28.248Z',
        pxObjClass: 'History-DXIL-DigV2-Work',
        PerformerID: 'User DigV2',
        pxInsName: 'DXIL-DIGV2-WORK C-7430!20230113T214628.248 GMT',
        pxLongitude: null,
        pzInsKey: 'HISTORY-DXIL-DIGV2-WORK DXIL-DIGV2-WORK C-7430!20230113T214628.248 GMT',
        MessageText: "Assigned to User DigV2 to 'complete task' ",
        pxHistoryForReference: 'DXIL-DIGV2-WORK C-7430',
        pxLatitude: null,
        pyMemo: null
      },
      {
        pxObjClass: 'History-DXIL-DigV2-Work',
        CreateDateTime: '2023-01-13T21:46:28.242Z',
        PerformerID: 'User DigV2',
        pxInsName: 'DXIL-DIGV2-WORK C-7430!20230113T214628.242 GMT',
        pxLongitude: null,
        pzInsKey: 'HISTORY-DXIL-DIGV2-WORK DXIL-DIGV2-WORK C-7430!20230113T214628.242 GMT',
        MessageText: "Assignment to 'complete task' completed by performing a 'EmbeddedData'.",
        pxHistoryForReference: 'DXIL-DIGV2-WORK C-7430',
        pyMemo: null,
        pxLatitude: null
      },
      {
        pxObjClass: 'History-DXIL-DigV2-Work',
        CreateDateTime: '2023-01-13T21:46:08.399Z',
        PerformerID: 'User DigV2',
        pxInsName: 'DXIL-DIGV2-WORK C-7430!20230113T214608.399 GMT',
        pxLongitude: null,
        pzInsKey: 'HISTORY-DXIL-DIGV2-WORK DXIL-DIGV2-WORK C-7430!20230113T214608.399 GMT',
        pxHistoryForReference: 'DXIL-DIGV2-WORK C-7430',
        MessageText: "Assigned to User DigV2 to 'complete task' ",
        pxLatitude: null,
        pyMemo: null
      },
      {
        CreateDateTime: '2023-01-13T21:46:08.393Z',
        pxObjClass: 'History-DXIL-DigV2-Work',
        PerformerID: 'User DigV2',
        pxInsName: 'DXIL-DIGV2-WORK C-7430!20230113T214608.393 GMT',
        pxLongitude: null,
        pzInsKey: 'HISTORY-DXIL-DIGV2-WORK DXIL-DIGV2-WORK C-7430!20230113T214608.393 GMT',
        MessageText: "Assignment to 'complete task' removed via ticket.",
        pxHistoryForReference: 'DXIL-DIGV2-WORK C-7430',
        pxLatitude: null,
        pyMemo: null
      },
      {
        pxObjClass: 'History-DXIL-DigV2-Work',
        CreateDateTime: '2023-01-13T21:46:07.584Z',
        PerformerID: 'User DigV2',
        pxInsName: 'DXIL-DIGV2-WORK C-7430!20230113T214607.584 GMT',
        pxLongitude: null,
        pzInsKey: 'HISTORY-DXIL-DIGV2-WORK DXIL-DIGV2-WORK C-7430!20230113T214607.584 GMT',
        pxHistoryForReference: 'DXIL-DIGV2-WORK C-7430',
        MessageText: "Assigned to User DigV2 to 'complete task' ",
        pyMemo: null,
        pxLatitude: null
      },
      {
        pxObjClass: 'History-DXIL-DigV2-Work',
        CreateDateTime: '2023-01-13T21:46:07.579Z',
        PerformerID: 'User DigV2',
        pxInsName: 'DXIL-DIGV2-WORK C-7430!20230113T214607.579 GMT',
        pzInsKey: 'HISTORY-DXIL-DIGV2-WORK DXIL-DIGV2-WORK C-7430!20230113T214607.579 GMT',
        pxLongitude: null,
        pxHistoryForReference: 'DXIL-DIGV2-WORK C-7430',
        MessageText: "Assignment to 'complete task' completed by performing a 'EmbeddedData'.",
        pxLatitude: null,
        pyMemo: null
      },
      {
        pxObjClass: 'History-DXIL-DigV2-Work',
        CreateDateTime: '2023-01-13T21:46:06.262Z',
        PerformerID: 'User DigV2',
        pxInsName: 'DXIL-DIGV2-WORK C-7430!20230113T214606.262 GMT',
        pxLongitude: null,
        pzInsKey: 'HISTORY-DXIL-DIGV2-WORK DXIL-DIGV2-WORK C-7430!20230113T214606.262 GMT',
        pxHistoryForReference: 'DXIL-DIGV2-WORK C-7430',
        MessageText: "Assigned to User DigV2 to 'complete task' ",
        pyMemo: null,
        pxLatitude: null
      },
      {
        pxObjClass: 'History-DXIL-DigV2-Work',
        CreateDateTime: '2023-01-13T21:46:06.256Z',
        PerformerID: 'User DigV2',
        pxInsName: 'DXIL-DIGV2-WORK C-7430!20230113T214606.256 GMT',
        pxLongitude: null,
        pzInsKey: 'HISTORY-DXIL-DIGV2-WORK DXIL-DIGV2-WORK C-7430!20230113T214606.256 GMT',
        pxHistoryForReference: 'DXIL-DIGV2-WORK C-7430',
        MessageText: "Assignment to 'complete task' removed via ticket.",
        pxLatitude: null,
        pyMemo: null
      },
      {
        pxObjClass: 'History-DXIL-DigV2-Work',
        CreateDateTime: '2023-01-13T21:45:57.503Z',
        PerformerID: 'User DigV2',
        pxInsName: 'DXIL-DIGV2-WORK C-7430!20230113T214557.503 GMT',
        pxLongitude: null,
        pzInsKey: 'HISTORY-DXIL-DIGV2-WORK DXIL-DIGV2-WORK C-7430!20230113T214557.503 GMT',
        MessageText: "Assigned to User DigV2 to 'complete task' ",
        pxHistoryForReference: 'DXIL-DIGV2-WORK C-7430',
        pxLatitude: null,
        pyMemo: null
      },
      {
        pxObjClass: 'History-DXIL-DigV2-Work',
        CreateDateTime: '2023-01-13T21:45:57.496Z',
        PerformerID: 'User DigV2',
        pxInsName: 'DXIL-DIGV2-WORK C-7430!20230113T214557.496 GMT',
        pxLongitude: null,
        pzInsKey: 'HISTORY-DXIL-DIGV2-WORK DXIL-DIGV2-WORK C-7430!20230113T214557.496 GMT',
        pxHistoryForReference: 'DXIL-DIGV2-WORK C-7430',
        MessageText: "Assignment to 'complete task' completed by performing a 'EmbeddedData'.",
        pyMemo: null,
        pxLatitude: null
      },
      {
        pxObjClass: 'History-DXIL-DigV2-Work',
        CreateDateTime: '2023-01-13T21:45:36.250Z',
        PerformerID: 'User DigV2',
        pxInsName: 'DXIL-DIGV2-WORK C-7430!20230113T214536.250 GMT',
        pxLongitude: null,
        pzInsKey: 'HISTORY-DXIL-DIGV2-WORK DXIL-DIGV2-WORK C-7430!20230113T214536.250 GMT',
        pxHistoryForReference: 'DXIL-DIGV2-WORK C-7430',
        MessageText: "Assigned to User DigV2 to 'complete task' ",
        pxLatitude: null,
        pyMemo: null
      },
      {
        pxObjClass: 'History-DXIL-DigV2-Work',
        CreateDateTime: '2023-01-13T21:45:36.244Z',
        PerformerID: 'User DigV2',
        pxInsName: 'DXIL-DIGV2-WORK C-7430!20230113T214536.244 GMT',
        pxLongitude: null,
        pzInsKey: 'HISTORY-DXIL-DIGV2-WORK DXIL-DIGV2-WORK C-7430!20230113T214536.244 GMT',
        pxHistoryForReference: 'DXIL-DIGV2-WORK C-7430',
        MessageText: "Assignment to 'complete task' removed via ticket.",
        pxLatitude: null,
        pyMemo: null
      },
      {
        CreateDateTime: '2023-01-13T21:45:35.364Z',
        pxObjClass: 'History-DXIL-DigV2-Work',
        PerformerID: 'User DigV2',
        pxInsName: 'DXIL-DIGV2-WORK C-7430!20230113T214535.364 GMT',
        pxLongitude: null,
        pzInsKey: 'HISTORY-DXIL-DIGV2-WORK DXIL-DIGV2-WORK C-7430!20230113T214535.364 GMT',
        pxHistoryForReference: 'DXIL-DIGV2-WORK C-7430',
        MessageText: "Assigned to User DigV2 to 'complete task' ",
        pxLatitude: null,
        pyMemo: null
      },
      {
        CreateDateTime: '2023-01-13T21:45:35.356Z',
        pxObjClass: 'History-DXIL-DigV2-Work',
        PerformerID: 'User DigV2',
        pxInsName: 'DXIL-DIGV2-WORK C-7430!20230113T214535.356 GMT',
        pzInsKey: 'HISTORY-DXIL-DIGV2-WORK DXIL-DIGV2-WORK C-7430!20230113T214535.356 GMT',
        pxLongitude: null,
        MessageText: "Assignment to 'complete task' completed by performing a 'EmbeddedData'.",
        pxHistoryForReference: 'DXIL-DIGV2-WORK C-7430',
        pyMemo: null,
        pxLatitude: null
      },
      {
        pxObjClass: 'History-DXIL-DigV2-Work',
        CreateDateTime: '2023-01-13T21:45:33.224Z',
        PerformerID: 'User DigV2',
        pxInsName: 'DXIL-DIGV2-WORK C-7430!20230113T214533.224 GMT',
        pxLongitude: null,
        pzInsKey: 'HISTORY-DXIL-DIGV2-WORK DXIL-DIGV2-WORK C-7430!20230113T214533.224 GMT',
        pxHistoryForReference: 'DXIL-DIGV2-WORK C-7430',
        MessageText: "Assigned to User DigV2 to 'complete task' ",
        pxLatitude: null,
        pyMemo: null
      },
      {
        CreateDateTime: '2023-01-13T21:45:33.218Z',
        pxObjClass: 'History-DXIL-DigV2-Work',
        PerformerID: 'User DigV2',
        pxInsName: 'DXIL-DIGV2-WORK C-7430!20230113T214533.218 GMT',
        pzInsKey: 'HISTORY-DXIL-DIGV2-WORK DXIL-DIGV2-WORK C-7430!20230113T214533.218 GMT',
        pxLongitude: null,
        pxHistoryForReference: 'DXIL-DIGV2-WORK C-7430',
        MessageText: "Assignment to 'complete task' removed via ticket.",
        pxLatitude: null,
        pyMemo: null
      },
      {
        CreateDateTime: '2023-01-13T21:45:32.436Z',
        pxObjClass: 'History-DXIL-DigV2-Work',
        PerformerID: 'User DigV2',
        pxInsName: 'DXIL-DIGV2-WORK C-7430!20230113T214532.436 GMT',
        pzInsKey: 'HISTORY-DXIL-DIGV2-WORK DXIL-DIGV2-WORK C-7430!20230113T214532.436 GMT',
        pxLongitude: null,
        pxHistoryForReference: 'DXIL-DIGV2-WORK C-7430',
        MessageText: "Assigned to User DigV2 to 'complete task' ",
        pxLatitude: null,
        pyMemo: null
      },
      {
        CreateDateTime: '2023-01-13T21:45:32.430Z',
        pxObjClass: 'History-DXIL-DigV2-Work',
        PerformerID: 'User DigV2',
        pxInsName: 'DXIL-DIGV2-WORK C-7430!20230113T214532.430 GMT',
        pxLongitude: null,
        pzInsKey: 'HISTORY-DXIL-DIGV2-WORK DXIL-DIGV2-WORK C-7430!20230113T214532.430 GMT',
        MessageText: "Assignment to 'complete task' completed by performing a 'EmbeddedData'.",
        pxHistoryForReference: 'DXIL-DIGV2-WORK C-7430',
        pxLatitude: null,
        pyMemo: null
      },
      {
        CreateDateTime: '2023-01-13T21:45:25.003Z',
        pxObjClass: 'History-DXIL-DigV2-Work',
        PerformerID: 'User DigV2',
        pxInsName: 'DXIL-DIGV2-WORK C-7430!20230113T214525.003 GMT',
        pzInsKey: 'HISTORY-DXIL-DIGV2-WORK DXIL-DIGV2-WORK C-7430!20230113T214525.003 GMT',
        pxLongitude: null,
        MessageText: "Assigned to User DigV2 to 'complete task' ",
        pxHistoryForReference: 'DXIL-DIGV2-WORK C-7430',
        pxLatitude: null,
        pyMemo: null
      },
      {
        pxObjClass: 'History-DXIL-DigV2-Work',
        CreateDateTime: '2023-01-13T21:45:24.994Z',
        PerformerID: 'User DigV2',
        pxInsName: 'DXIL-DIGV2-WORK C-7430!20230113T214524.994 GMT',
        pxLongitude: null,
        pzInsKey: 'HISTORY-DXIL-DIGV2-WORK DXIL-DIGV2-WORK C-7430!20230113T214524.994 GMT',
        pxHistoryForReference: 'DXIL-DIGV2-WORK C-7430',
        MessageText: "Assignment to 'complete task' completed by performing a 'Select Test'.",
        pyMemo: null,
        pxLatitude: null
      },
      {
        pxObjClass: 'History-DXIL-DigV2-Work',
        CreateDateTime: '2023-01-13T21:45:23.570Z',
        PerformerID: 'User DigV2',
        pxInsName: 'DXIL-DIGV2-WORK C-7430!20230113T214523.570 GMT',
        pzInsKey: 'HISTORY-DXIL-DIGV2-WORK DXIL-DIGV2-WORK C-7430!20230113T214523.570 GMT',
        pxLongitude: null,
        pxHistoryForReference: 'DXIL-DIGV2-WORK C-7430',
        MessageText: "Assigned to User DigV2 to 'complete task' ",
        pyMemo: null,
        pxLatitude: null
      },
      {
        pxObjClass: 'History-DXIL-DigV2-Work',
        CreateDateTime: '2023-01-13T21:45:23.563Z',
        PerformerID: 'User DigV2',
        pxInsName: 'DXIL-DIGV2-WORK C-7430!20230113T214523.563 GMT',
        pzInsKey: 'HISTORY-DXIL-DIGV2-WORK DXIL-DIGV2-WORK C-7430!20230113T214523.563 GMT',
        pxLongitude: null,
        MessageText: 'ChangeStageAutomatic\tCreate\tPick Test',
        pxHistoryForReference: 'DXIL-DIGV2-WORK C-7430',
        pxLatitude: null,
        pyMemo: 'ChangeStageAutomatic\tCreate\tPick Test'
      },
      {
        pxObjClass: 'History-DXIL-DigV2-Work',
        CreateDateTime: '2023-01-13T21:45:23.550Z',
        PerformerID: 'User DigV2',
        pxInsName: 'DXIL-DIGV2-WORK C-7430!20230113T214523.550 GMT',
        pzInsKey: 'HISTORY-DXIL-DIGV2-WORK DXIL-DIGV2-WORK C-7430!20230113T214523.550 GMT',
        pxLongitude: null,
        MessageText: 'Item created.',
        pxHistoryForReference: 'DXIL-DIGV2-WORK C-7430',
        pyMemo: null,
        pxLatitude: null
      }
    ],
    hasMoreResults: false
  },
  status: 200,
  statusText: '',
  headers: {
    'content-length': '1328',
    'content-type': 'application/json;charset=UTF-8'
  },
  request: {}
};
export default historyData;
