// @pega/cosmos-react-core/lib/components/AppShell/AppShell.styles.js
import styled, { css } from 'styled-components';
import { darken, getContrast, lighten, readableColor, transparentize } from 'polished';
import { useContext } from 'react';
import { StyledAvatar } from '../Avatar';
import { StyledIcon } from '../Icon';
import { StyledImage } from '../Image/Image';
import { defaultThemeProp } from '../../theme';
import { tryCatch } from '../../utils';
import { calculateFontSize } from '../../styles';
import { StyledMenuItem } from '../Menu/Menu.styles';
import { useDirection } from '../../hooks';
import { StyledBareButton } from '../Button/BareButton';
import Button from '../Button';
import { StyledTooltip } from '../Tooltip';
import { StyledProgressBackdrop } from '../Progress';
import AppShellContext from './AppShellContext';
import { navContrastColors, pxToRem } from './style-utils';
export const navWidth = '4rem';
export const navOpenWidth = '18.75rem';
export const headerHeight = '3rem';
export const StyledAppInfo = styled.a(({ theme }) => {
    const fontSize = calculateFontSize(theme.base['font-size'], theme.base['font-scale']);
    const { hoverBg, foreground } = navContrastColors(theme);
    return css `
    padding: ${theme.base.spacing} calc(${navWidth} / 4);
    grid-template-columns: calc(${navWidth} / 2) auto;
    column-gap: calc(${navWidth} / 4);

    &:hover,
    button&:hover {
      background-color: ${hoverBg};
      text-decoration: none;
    }

    &:focus:not([disabled]) {
      outline: none;
      box-shadow: ${theme.base.shadow['focus-inset']};
    }

    ${StyledImage} {
      display: inline-block;
      width: calc(${navWidth} / 2);
    }

    span {
      overflow: hidden;
      text-overflow: ellipsis;
    }

    span:last-of-type {
      font-size: ${fontSize.xs};
      color: ${foreground};
    }

    span:first-of-type {
      font-size: ${fontSize.s};
      color: ${foreground};
    }
  `;
});
StyledAppInfo.defaultProps = defaultThemeProp;
export const StyledScrollWrap = styled.div(() => {
    return css `
    flex-grow: 1;
    overflow-x: hidden;
    overflow-y: auto;
  `;
});
StyledScrollWrap.defaultProps = defaultThemeProp;
export const StyledCountIcon = styled.span(({ theme: { base: { spacing } } }) => {
    return css `
      position: absolute;
      inset-inline-start: calc(4 * ${spacing});
      top: calc(0.5 * ${spacing});
      padding: 0 calc(0.5 * ${spacing});
    `;
});
StyledCountIcon.defaultProps = defaultThemeProp;
export const StyledSingleSelectIconBox = styled.div(({ theme }) => {
    return css `
    height: ${theme.base['hit-area']['mouse-min']};
    width: ${theme.base['hit-area']['mouse-min']};
    margin-inline-start: calc(2 * ${theme.base.spacing});
  `;
});
StyledSingleSelectIconBox.defaultProps = defaultThemeProp;
export const StyledNavListItemExpandCollapse = styled.svg ``;
export const StyledUtilIconCount = styled.span `
  flex-shrink: 0;
`;
export const StyledNavItemIconBox = styled.div(({ theme }) => {
    return css `
    > ${StyledIcon}, ${StyledUtilIconCount} > ${StyledIcon} {
      margin-inline: calc((${navWidth} - 1.125rem) / 2);
    }

    ${StyledAvatar} {
      margin-inline: calc((${navWidth} - ${theme.base['hit-area']['mouse-min']}) / 2);
      font-weight: ${theme.base['font-weight'].normal};
    }
  `;
});
StyledNavItemIconBox.defaultProps = defaultThemeProp;
export const StyledAppNav = styled.nav(({ appHeader, showEnv, theme }) => {
    const navBg = theme.components['app-shell'].nav['background-color'];
    const { navOpen } = useContext(AppShellContext);
    return css `
      position: fixed;
      z-index: calc(${theme.base['z-index'].backdrop} - 2);
      background: ${navBg};
      max-width: calc(100vw - ${theme.base['hit-area']['mouse-min']});
      width: ${navOpen ? navOpenWidth : navWidth};
      height: calc(100vh - ${appHeader ? headerHeight : '0rem'});
      overflow: hidden;
      white-space: nowrap;
      transition: width ${`${theme.base.animation.speed} ${theme.base.animation.timing.ease}`};
      padding-block-end: ${theme.base.spacing};
      border-inline-end: ${pxToRem(1)} solid ${transparentize(0.9, '#000')};
      ${!showEnv &&
        css `
        padding-block-start: ${theme.base.spacing};
      `}
      @media (pointer: coarse) {
        max-width: calc(100vw - ${theme.base['hit-area']['finger-min']});
      }

      /* App navigation can be focused with shortcut */
      :focus-visible {
        box-shadow: ${theme.base.shadow['focus-inset']};
      }

      ${theme.base['custom-scrollbar'] &&
        css `
        * {
          scrollbar-color: rgba(255, 255, 255, 0.6) rgba(0, 0, 0, 0.2);
          scrollbar-width: thin;
        }

        @supports not ((scrollbar-width: thin) or (scrollbar-color: black white)) {
          /* stylelint-disable unit-allowed-list */

          *::-webkit-scrollbar {
            width: 12px;
            height: 12px;
          }

          *::-webkit-scrollbar-track {
            background-color: rgba(0, 0, 0, 0.2);
          }

          *::-webkit-scrollbar-corner {
            background-color: rgba(0, 0, 0, 0.2);
            border-bottom-right-radius: inherit;
          }

          *::-webkit-scrollbar-thumb {
            background-color: transparent;
            border: 3px solid transparent;
            border-radius: calc(1.125 * ${theme.base['border-radius']});
            box-shadow: inset 0 0 2px 4px rgba(255, 255, 255, 0.6);
          }

          *::-webkit-scrollbar-thumb:hover {
            box-shadow: inset 0 0 2px 4px rgba(255, 255, 255, 0.7);
          }

          *::-webkit-scrollbar-track:horizontal {
            border-bottom-left-radius: inherit;
          }

          *::-webkit-scrollbar-track:vertical {
            border-top-right-radius: inherit;
          }

          /* stylelint-disable-next-line selector-pseudo-class-no-unknown */
          *::-webkit-scrollbar-track:not(:corner-present) {
            border-bottom-right-radius: inherit;
          }

          /* stylelint-enable unit-allowed-list */
        }
      `}
    `;
});
StyledAppNav.defaultProps = defaultThemeProp;
export const StyledCaseClose = styled.button ``;
export const StyledNavListItem = styled.li(({ theme, nestedListCollapsed, singleSelect }) => {
    const { base: { 'font-weight': { 'semi-bold': semiBold } } } = theme;
    const { navOpen } = useContext(AppShellContext);
    const { hoverBg, foreground, border: borderColor } = navContrastColors(theme);
    const { rtl } = useDirection();
    return css `
    display: block;
    position: relative;

    ${StyledIcon} {
      font-size: 1.25rem;
    }

    ${StyledCaseClose} {
      display: flex;
      position: absolute;
      opacity: 0;
      left: ${navOpen ? 'auto' : '100%'};
      top: 50%;
      bottom: 50%;
      margin: auto 0;
      color: ${foreground};
      transition-property: transform, opacity;
      transition-duration: ${theme.base.animation.speed};
      transition-timing-function: ${theme.base.animation.timing.ease};
      ${navOpen &&
        css `
        inset-inline-end: 1rem;

        &:hover,
        &:focus {
          opacity: 1;
        }
      `}
    }

    > :first-child {
      position: relative;
      display: flex;
      align-items: center;
      width: 100%;
      padding: calc(1.25 * ${theme.base.spacing}) 0;
      white-space: nowrap;
      color: ${foreground};
      cursor: pointer;
      text-decoration: none;
      border-block: ${pxToRem(1)} solid transparent;

      span:not(${StyledCountIcon}) {
        overflow: hidden;
        text-overflow: ellipsis;
        color: inherit;
        white-space: nowrap;
      }

      &:hover,
      &:focus {
        color: ${foreground};
        background: ${hoverBg};
      }

      &:hover + ${StyledCaseClose}, &:focus + ${StyledCaseClose} {
        ${navOpen &&
        css `
          opacity: 1;
        `}
      }

      &:focus {
        outline: none;
        box-shadow: ${theme.base.shadow['focus-inset']};
      }

      ${StyledNavListItemExpandCollapse} {
        position: absolute;
        left: ${navOpen ? 'auto' : '100%'};
        ${navOpen &&
        css `
          inset-inline-end: 1rem;
        `}
        margin: 0;
        opacity: ${navOpen ? '1' : '0'};
        transition-property: transform, opacity;
        transition-duration: ${theme.base.animation.speed};
        transition-timing-function: ${theme.base.animation.timing.ease};

        ${!nestedListCollapsed &&
        css `
          transform: rotateZ(${rtl ? '-90deg' : '90deg'});
        `}
      }

      ${singleSelect &&
        css `
        ${StyledNavItemIconBox} > ${StyledIcon} {
          margin-inline: ${theme.base.spacing};
        }
      `}

      ${StyledSingleSelectIconBox} + span {
        margin-inline-start: ${theme.base.spacing};
      }
    }

    [aria-current='page'] {
      position: relative;
      font-weight: ${semiBold};
      border-block: ${pxToRem(1)} solid ${borderColor};

      &::before {
        content: '';
        position: absolute;
        width: ${pxToRem(4)};
        inset-inline-start: 0;
        top: 0;
        bottom: 0;
        background-color: ${foreground};
      }
    }
  `;
});
StyledNavListItem.defaultProps = defaultThemeProp;
export const StyledNavList = styled.ul ``;
export const StyledNavCasesList = styled.div(({ theme }) => {
    return css `
    ${StyledNavListItem} {
      > :first-child {
        padding: ${theme.base.spacing} 0;
      }
    }
  `;
});
StyledNavCasesList.defaultProps = defaultThemeProp;
export const StyledNestedNavList = styled(StyledNavList)(({ theme }) => {
    return css `
    ${StyledNavListItem} {
      > :first-child {
        padding-inline-start: calc(${navWidth} + 2 * ${theme.base.spacing});
        font-weight: ${theme.base['font-weight'].normal};
      }
    }
  `;
});
StyledNestedNavList.defaultProps = defaultThemeProp;
export const StyledCaseTypes = styled.div(props => {
    const { theme: { base: { animation: { speed, timing: { ease: acceleration } } } } } = props;
    return css `
    button {
      ${StyledNavItemIconBox} {
        transition: transform ${speed} ${acceleration};
      }

      &[aria-expanded='false'] ${StyledNavItemIconBox} {
        transform: none;
      }

      &[aria-expanded='true'] ${StyledNavItemIconBox} {
        transform: rotateZ(45deg);
      }
    }
  `;
});
StyledCaseTypes.defaultProps = defaultThemeProp;
export const StyledUtils = styled.div(({ theme }) => {
    const { border: borderColor } = navContrastColors(theme);
    return css `
    border-top: 0.0625rem solid ${borderColor};

    ${StyledBareButton} > ${StyledIcon}:first-child, ${StyledBareButton} > ${StyledUtilIconCount} {
      margin: 0 calc((${navWidth} - 1.125rem) / 2);
    }
  `;
});
StyledUtils.defaultProps = defaultThemeProp;
export const StyledMobileNav = styled.header(({ theme }) => {
    const mobileNavColor = theme.base.palette['brand-primary'];
    const contrastColor = tryCatch(() => readableColor(mobileNavColor));
    return css `
    min-width: 100%;
    min-height: ${headerHeight};
    z-index: ${theme.base['z-index'].drawer};
    background-color: ${mobileNavColor};
    color: ${contrastColor};
    position: sticky;
    top: 0;
    padding-inline-start: ${theme.base.spacing};

    @media (pointer: coarse) {
      padding-inline-start: 0;
    }
  `;
});
StyledMobileNav.defaultProps = defaultThemeProp;
export const StyledLoading = styled.div `
  font-size: 2rem;
`;
export const StyledAppMain = styled.main(({ appHeader, mobileNav, navOpen, headerOffset = 0, theme }) => {
    const heightOffset = appHeader || mobileNav ? headerHeight : '0rem';
    return css `
    --appshell-offset: ${headerOffset}px;
    /* stylelint-disable-next-line length-zero-no-unit */
    --appshell-horizontal-offset: 0rem;

    @media (min-width: ${theme.base.breakpoints.sm}) {
      --appshell-horizontal-offset: ${navOpen ? navOpenWidth : navWidth};
    }
    position: relative;
    margin-inline-start: ${mobileNav ? 0 : navWidth};
    min-height: calc(100vh - ${heightOffset});
    background-color: ${theme.base.palette['app-background']};

    :focus {
      outline: none;
    }

    ${navOpen &&
        !mobileNav &&
        css `
      margin-inline-start: ${navOpenWidth};
    `}

    ${!mobileNav &&
        css `
      transition: margin-inline-start
        ${`${theme.base.animation.speed} ${theme.base.animation.timing.ease}`};
    `}

    & > ${StyledProgressBackdrop} {
      max-height: calc(100vh - ${heightOffset});
      max-width: calc(100vw - ${navOpen ? navOpenWidth : navWidth});
      inset-block-start: ${heightOffset};
      inset-inline-start: ${navOpen ? navOpenWidth : navWidth};
      position: fixed;
      transition-property: opacity, max-height, max-width, inset-block-start, inset-inline-start;
      transition-duration: ${theme.base.animation.speed};
      transition-timing-function: ${theme.base.animation.timing.ease};
    }
  `;
});
StyledAppMain.defaultProps = defaultThemeProp;
export const StyledBannerRegion = styled.div(({ theme }) => {
    return css `
    padding: calc(2 * ${theme.base.spacing});
  `;
});
StyledBannerRegion.defaultProps = defaultThemeProp;
export const StyledAppShellTooltip = styled(StyledTooltip)(({ theme }) => {
    const bg = theme.components.tooltip['background-color'];
    const color = readableColor(bg);
    return css `
    background-color: ${bg};
    padding: 0;
    header {
      color: ${color};
      border-bottom: none;
    }
    ${StyledMenuItem} {
      background-color: ${bg};
      color: ${color};

      &:hover,
      &:focus {
        background-color: ${bg};
        box-shadow: ${theme.base.shadow.focus};
      }
    }
  `;
});
StyledAppShellTooltip.defaultProps = defaultThemeProp;
export const StyledAppShellToggleButton = styled(Button)(({ theme }) => {
    const navBg = theme.components['app-shell'].nav['background-color'];
    let bg = darken(0.2, navBg);
    if (getContrast(navBg, bg) < 1.3) {
        bg = lighten(0.2, navBg);
    }
    const color = readableColor(bg);
    return css `
    background-color: ${bg};
    color: ${color};
    border: none;
    margin-top: calc(1 * ${theme.base.spacing});

    ${StyledIcon} {
      width: 1.25rem;
      height: 1.25rem;
      font-weight: bold;
    }

    &:enabled:focus,
    :not([disabled]):focus {
      box-shadow: ${theme.base.shadow.focus};
    }
  `;
});
StyledAppShellToggleButton.defaultProps = defaultThemeProp;
