// @pega/cosmos-react-rte/lib/components/Editor/Editor.styles.js
import styled, { css } from 'styled-components';
import { defaultThemeProp } from '@pega/cosmos-react-core';
export const offscreenSelectionStyles = `
  .mce-offscreen-selection {
    border: 0;
    clip: rect(0 0 0 0);
    height: 1px;
    margin: -1px;
    overflow: hidden;
    padding: 0;
    position: absolute;
    whiteSpace: nowrap;
    width: 1px;
  }
`;
export const mceContentBodyStyles = `
  .mce-content-body {
    overflow-wrap: break-word;
  }
`;
export const placeholderStyles = `
  .mce-content-body[data-mce-placeholder] {
    position: relative;
  }
  .mce-content-body[data-mce-placeholder]:not(.mce-visualblocks)::before {
    cursor: text;
    color: rgb(117, 117, 117);
    content: attr(data-mce-placeholder);
    position: absolute;
  }
`;
export const StyledEditorContainer = styled.div(({ theme }) => {
    return css `
    &:hover {
      cursor: text;
    }

    iframe {
      display: block;
      width: 100%;
      height: ${theme.components['text-area']['min-height']};
    }

    .tox-statusbar {
      display: none !important;
    }
  `;
});
StyledEditorContainer.defaultProps = defaultThemeProp;
