// @pega/cosmos-react-work/lib/components/Assignments/Assignments.styles.js
import styled, { css } from 'styled-components';
import { Button, Card, CardContent, defaultThemeProp, StyledCard, StyledIcon, useDirection } from '@pega/cosmos-react-core';
export const StyledExpandButton = styled(Button)(({ theme: { base: { animation } } }) => {
    const { rtl } = useDirection();
    return css `
      ${StyledIcon} {
        transition: transform ${animation.speed} ${animation.timing.ease};
      }

      &[aria-expanded='true'] ${StyledIcon} {
        transform: rotate(90deg);
      }

      &[aria-expanded='false'] ${StyledIcon} {
        transform: rotate(${rtl ? 180 : 0}deg);
      }

      &:hover,
      &:active {
        text-decoration: none;
      }
    `;
});
StyledExpandButton.defaultProps = defaultThemeProp;
export const StyledAssignments = styled(Card)(({ theme, openItem }) => {
    return openItem
        ? css `
        box-shadow: ${theme.base.shadow.low};
      `
        : null;
});
StyledAssignments.defaultProps = defaultThemeProp;
export const StyledCardContent = styled(CardContent)(({ theme, openItem }) => {
    return css `
      &:not(${StyledCard} ${StyledCard} > &) {
        padding: ${openItem ? theme.base.spacing : '0'} calc(2 * ${theme.base.spacing});
      }
    `;
});
StyledCardContent.defaultProps = defaultThemeProp;
