// @pega/cosmos-react-core/lib/components/DateTime/Input/MonthInput.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useMemo, forwardRef, useEffect, useRef } from 'react';
import styled from 'styled-components';
import { defaultThemeProp } from '../../../theme';
import { range, withTestIds } from '../../../utils';
import { useConfiguration, useConsolidatedRef, useFocusWithin, useTestIds } from '../../../hooks';
import FormField from '../../FormField';
import Select, { Option } from '../../Select';
import { StyledSelect } from '../../Select/Select';
import Flex from '../../Flex';
import { StyledFormControl } from '../../FormControl';
import { getMonthInputTestIds } from '../DateTime.test-ids';
import { parseToDate, getLocalizedMonths } from './utils';
import StyledDateTime from './DateTime.styles';
export const StyledMonthInput = styled(StyledDateTime) `
  padding: 0;
  border: 0;

  ${StyledSelect} {
    min-width: 0;

    &:focus {
      z-index: 1;
    }

    &:first-child {
      border-start-end-radius: 0;
      border-end-end-radius: 0;
    }

    &:last-child {
      max-width: max-content;
      margin-inline-start: -0.0625rem;
      border-start-start-radius: 0;
      border-end-start-radius: 0;
    }
  }
`;
StyledMonthInput.defaultProps = defaultThemeProp;
const convertToCallbackParameter = (date, { month, year }) => {
    if ([month, year].every(Boolean)) {
        const resultDate = new Date(date);
        resultDate.setUTCFullYear(Number(year), Number(month));
        return {
            valueAsISOString: resultDate.toISOString(),
            valueAsTimestamp: resultDate.getTime()
        };
    }
    if ([year, month].every(p => !p)) {
        return {
            valueAsISOString: undefined,
            valueAsTimestamp: undefined
        };
    }
    return {
        valueAsISOString: '',
        valueAsTimestamp: NaN,
        state: 'incomplete'
    };
};
const MonthInput = forwardRef(function MonthInput(props, ref) {
    const { locale } = useConfiguration();
    const { testId, value, min, max, id, label, labelHidden, info, status, required, readOnly, disabled, onChange, onFocus, onBlur, autoFocus, additionalInfo, ...restProps } = props;
    const testIds = useTestIds(testId, getMonthInputTestIds);
    const date = value ? parseToDate(value) : undefined;
    const dateToGetFullYearFrom = date ?? new Date();
    const minDate = parseToDate(min ??
        new Date(dateToGetFullYearFrom).setUTCFullYear(dateToGetFullYearFrom.getUTCFullYear() - 10));
    const maxDate = parseToDate(max ??
        new Date(dateToGetFullYearFrom).setUTCFullYear(dateToGetFullYearFrom.getUTCFullYear() + 10));
    minDate.setUTCMinutes(minDate.getUTCMinutes() - minDate.getTimezoneOffset());
    maxDate.setUTCMinutes(maxDate.getUTCMinutes() - maxDate.getTimezoneOffset());
    const minYear = minDate.getUTCFullYear();
    const maxYear = maxDate.getUTCFullYear();
    const years = range(minYear, maxYear);
    const months = useMemo(() => getLocalizedMonths(locale), [locale]);
    const monthSelectRef = useRef(null);
    const yearSelectRef = useRef(null);
    useEffect(() => {
        if (monthSelectRef.current && yearSelectRef.current) {
            monthSelectRef.current.value = date?.getUTCMonth().toString() ?? '';
            yearSelectRef.current.value = date?.getUTCFullYear().toString() ?? '';
        }
    }, [date?.toDateString()]);
    const pickParts = () => {
        return {
            month: monthSelectRef.current?.value || '',
            year: yearSelectRef.current?.value || ''
        };
    };
    const onFocusChange = (focused) => {
        const callbackParam = convertToCallbackParameter(date ? new Date(date) : new Date(), pickParts());
        if (onFocus && focused)
            onFocus(callbackParam);
        else if (onBlur && !focused)
            onBlur(callbackParam);
    };
    const onSelectChange = () => {
        onChange?.(convertToCallbackParameter(date ? new Date(date) : new Date(), pickParts()));
    };
    const containerRef = useConsolidatedRef(ref);
    useFocusWithin([containerRef], onFocusChange);
    const isMonthOptionDisabled = (monthIndex) => {
        const year = yearSelectRef.current && yearSelectRef.current.value
            ? parseInt(yearSelectRef.current.value, 10)
            : new Date().getUTCFullYear();
        return ((year === minDate.getUTCFullYear() && monthIndex < minDate.getUTCMonth()) ||
            (year === maxDate.getUTCFullYear() && monthIndex > maxDate.getUTCMonth()));
    };
    const monthOptions = months.map((month, index) => (_jsx(Option, { disabled: isMonthOptionDisabled(index), value: index.toString(), children: months[index] }, month)));
    const yearOptions = years.map(year => (_jsx(Option, { value: year.toString(), children: year.toString() }, year.toString())));
    const displayNames = new Intl.DisplayNames(locale, { style: 'long', type: 'dateTimeField' });
    const Month = (_jsxs(Select, { "data-testid": testIds.controlMonth, ref: monthSelectRef, "aria-label": displayNames.of('month'), readOnly: readOnly, required: required, onChange: onSelectChange, status: status, disabled: disabled, autoFocus: autoFocus, children: [!required && _jsx(Option, { children: " " }, 'null'), monthOptions] }, 'month'));
    const Year = (_jsxs(Select, { "data-testid": testIds.controlYear, ref: yearSelectRef, "aria-label": displayNames.of('year'), readOnly: readOnly, required: required, onChange: onSelectChange, status: status, disabled: disabled, children: [!required && _jsx(Option, { children: " " }, 'null'), yearOptions] }, 'year'));
    const Comp = (_jsxs(Flex, { "data-testid": testIds.root, as: StyledMonthInput, forwardedAs: StyledFormControl, container: { alignItems: 'center', wrap: 'nowrap' }, ref: containerRef, status: status, disabled: disabled, readOnly: readOnly, ...restProps, children: [Month, Year] }));
    return label ? (_jsx(FormField, { as: 'fieldset', labelAs: 'legend', ...{
            testId: testIds,
            label,
            labelHidden,
            id,
            info,
            status,
            required,
            disabled,
            additionalInfo
        }, children: Comp })) : (Comp);
});
export default withTestIds(MonthInput, getMonthInputTestIds);
