// @pega/custom-dx-components/tests/assets/components/createAll/Launchpad/Pega_DXIL_MyPageWidget/index.tsx
import { withConfiguration } from '@pega/cosmos-react-core';
import { AppAnnouncement as PegaAppAnnouncement } from '@pega/cosmos-react-work';
import type { PConnFieldProps } from './PConnProps';

import StyledPegaDxilMyPageWidgetWrapper from './styles';

// interface for props
interface PegaDxilMyPageWidgetProps extends PConnFieldProps {
  // If any, enter additional props that only exist on TextInput here
    datasource: Array<any>;
    header: string;
    description: string;
    whatsnewlink: string;
    image: string;
}


// Duplicated runtime code from Constellation Design System Component

// props passed in combination of props from property panel (config.json) and run time props from Constellation
// any default values in config.pros should be set in defaultProps at bottom of this file
function PegaDxilMyPageWidget(props: PegaDxilMyPageWidgetProps) {

  const { header, description, datasource = [], whatsnewlink, image } = props;
  let details = [];
  // @ts-ignore
  if (datasource && datasource.source) {
    // @ts-ignore
    details = datasource.source.map((item) => {
      return item.name;
    });
  }
  return (
    <StyledPegaDxilMyPageWidgetWrapper>
    <PegaAppAnnouncement
      heading={header}
      description={description}
      details={details}
      whatsNewLink={whatsnewlink}
      image={image.replace(/ /g, '+')}
    />
    </StyledPegaDxilMyPageWidgetWrapper>
  );

}

export default withConfiguration(PegaDxilMyPageWidget);
