// @pega/commitlint-config/index.js
'use strict';

const plugin = require('@pega/commitlint-plugin');

const agileStudioPrefixes = ['EPIC-', 'US-', 'BUG-', 'ISSUE-', 'TASK-'];
const openAgilePrefixes = ['BUG-'];

const /** @type {import('@commitlint/types').UserConfig} */ Configuration = {
    parserPreset: {
      name: '',
      path: '',
      parserOpts: {
        headerPattern: /^([a-z]+)(?:\(([^)]+)\))?!?: (.+)$/,
        headerCorrespondence: ['type', 'scope', 'subject'],
        referenceActions: ['AgileStudio:'],
        issuePrefixes: [...agileStudioPrefixes, ...openAgilePrefixes],
        issuePrefixesCaseSensitive: true,
        noteKeywords: ['BREAKING CHANGE', 'AgileStudio', 'OpenAgile'],
        // cSpell:disable
        mergePattern: /^[Mm]erge/,
        revertPattern: /^[Rr]evert/
        // cSpell:enable
      }
    },

    defaultIgnores: false,

    rules: {
      'type-empty': [2, 'never'],
      'type-case': [2, 'always', 'lower-case'],
      'type-enum': [2, 'always', ['feature', 'fix', 'refactor', 'style', 'test', 'docs', 'pkg']],

      'subject-empty': [2, 'never'],
      'subject-case': [2, 'always', 'sentence-case'],
      'subject-full-stop': [2, 'always'],

      '@pega/header-reference': [2, 'never'],

      'body-max-length': [2, 'never', 0],

      'footer-leading-blank': [2, 'always'],

      '@pega/reference-action-enum': [
        2,
        'always',
        {
          actions: ['AgileStudio:'],
          caseSensitive: true
        }
      ],
      '@pega/reference-order': [2, 'always', { prefixes: agileStudioPrefixes }],

      '@pega/agile-studio': [2, 'always', { prefixes: agileStudioPrefixes }],
      '@pega/open-agile': [2, 'always', { prefixes: openAgilePrefixes }],

      '@pega/spellcheck': [2, 'always', { ignorePrefixes: ['OpenAgile:'] }],

      '@pega/merge-commit': [2, 'never'],
      '@pega/revert-commit': [2, 'never']
    },

    plugins: [plugin]
  };

module.exports = Configuration;
