// shortcuts/dist/utils.js
/* MAIN */
const Utils = {
    /* API */
    isArray: (value) => {
        return Array.isArray(value);
    },
    isEqual: (a, b) => {
        if (a.length !== b.length)
            return false;
        for (let i = 0, l = a.length; i < l; i++) {
            if (a[i] !== b[i])
                return false;
        }
        return true;
    },
    isFalsy: (value) => {
        return !value;
    },
    isKeyboardEvent: (event) => {
        return event.type.startsWith('key');
    },
    isMac: () => {
        if (typeof navigator !== 'object')
            return false;
        return /mac|ipod|iphone|ipad/i.test(navigator.platform);
    },
    isTruthy: (value) => {
        return !!value;
    },
    memoize: (fn) => {
        const cache = new Map();
        return (arg) => {
            const cached = cache.get(arg);
            if (cached || cache.has(arg))
                return cached; //TSC
            const result = fn(arg);
            cache.set(arg, result);
            return result;
        };
    }
};
/* EXPORT */
export default Utils;
