// @pega/custom-dx-components/tests/assets/components/createAll/Constellation/Pega_DXIL_MyCurrency/demo.stories.tsx

/* eslint-disable react/jsx-no-useless-fragment */
// @ts-nocheck
import type { Meta, StoryObj } from '@storybook/react';

import { configProps, fieldMetadata, stateProps } from './mock';

import PegaDxilMyCurrency from './index';

const meta: Meta<typeof PegaDxilMyCurrency> = {
  title: 'PegaDxilMyCurrency',
  component: PegaDxilMyCurrency,
  excludeStories: /.*Data$/
};

export default meta;
type Story = StoryObj<typeof PegaDxilMyCurrency>;

export const BasePegaDxilMyCurrency: Story = args => {

  const props = {
    value: Number(configProps.value),
    placeholder: configProps.placeholder,
    label: configProps.label,
    helperText: configProps.helperText,
    showGroupSeparators: configProps.showGroupSeparators,
    allowDecimals: configProps.allowDecimals,
    currencyISOCode: configProps.currencyISOCode,
    alwaysShowISOCode: configProps.alwaysShowISOCode,
    hasSuggestions: configProps.hasSuggestions,
    testId: configProps.testId,
    fieldMetadata,
    validatemessage: '',
    getPConnect: () => {
      return {
        getStateProps: () => {
          return stateProps;
        },
        getActionsApi: () => {
          return {
            updateFieldValue: () => {/* nothing */},
            triggerFieldChange: () => {/* nothing */}
          };
        },
        ignoreSuggestion: () => {/* nothing */},
        acceptSuggestion: () => {/* nothing */},
        setInheritedProps: () => {/* nothing */},
        resolveConfigProps: () => {/* nothing */},
        clearErrorMessages: () => {/* nothing */}
      };
    }
  };

  return (
    <>
      <PegaDxilMyCurrency {...props} {...args} />
    </>
  );
};

BasePegaDxilMyCurrency.args = {
  label: configProps.label,
  helperText: configProps.helperText,
  showGroupSeparators: configProps.showGroupSeparators,
  allowDecimals: configProps.allowDecimals,
  currencyISOCode: configProps.currencyISOCode,
  alwaysShowISOCode: configProps.alwaysShowISOCode,
  testId: configProps.testId,
  readOnly: configProps.readOnly,
  disabled: configProps.disabled,
  required: configProps.required,
  status: configProps.status,
  hideLabel: configProps.hideLabel,
  displayMode: configProps.displayMode,
  validatemessage: configProps.validatemessage
};
