// @pega/cosmos-react-work/lib/components/CaseView/UtilitiesSummary.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import styled, { css } from 'styled-components';
import { Card, Count, Flex, Icon, useI18n, VisuallyHiddenText } from '@pega/cosmos-react-core';
export const StyledUtilitiesSummary = styled.div(({ theme }) => {
    return css `
    width: 2.75rem;
    padding: ${theme.base.spacing} 0;
  `;
});
const UtilitiesSummary = ({ items = [] }) => {
    const t = useI18n();
    return items.length ? (_jsx(Card, { container: { direction: 'column', alignItems: 'center', gap: 3 }, as: StyledUtilitiesSummary, "aria-label": t('utilities_summary'), role: 'list', children: items.map(item => {
            return (_jsxs(Flex, { container: {
                    direction: 'column',
                    alignItems: 'center',
                    gap: 0.5
                }, item: { grow: 1 }, children: [_jsx(Icon, { name: item.iconName }), item.count !== undefined && _jsx(Count, { "aria-hidden": true, children: item.count }), _jsx(VisuallyHiddenText, { role: 'listitem', children: `${item.count ?? ''} ${item.name}` })] }, item.iconName));
        }) })) : null;
};
export default UtilitiesSummary;
