// @pega/custom-dx-components/tests/assets/components/createAll/Launchpad/Pega_DXIL_MyBoolean/demo.stories.tsx

/* eslint-disable react/jsx-no-useless-fragment */
// @ts-nocheck
import { useState } from 'react';
import type { Meta, StoryObj } from '@storybook/react';

import { configProps, stateProps } from './mock';

import PegaDxilMyBoolean from './index';

const meta: Meta<typeof PegaDxilMyBoolean> = {
  title: 'PegaDxilMyBoolean',
  component: PegaDxilMyBoolean,
  excludeStories: /.*Data$/
};

export default meta;
type Story = StoryObj<typeof PegaDxilMyBoolean>;

export const BasePegaDxilMyBoolean: Story = args => {

  const [value, setValue] = useState(configProps.value);

  const props = {
    value,
    label: configProps.label,
    helperText: configProps.helperText,
    caption: configProps.caption,
    validatemessage: configProps.validatemessage,
    hideLabel: configProps.hideLabel,
    testId: configProps.testId,
    additionalProps: configProps.additionalProps,
    displayMode: configProps.displayMode,
    variant: configProps.variant,
    trueLabel: configProps.trueLabel,
    falseLabel: configProps.falseLabel,
    disabled: configProps.disabled,
    readOnly: configProps.readOnly,
    required: configProps.required,
    getPConnect: () => {
      return {
        getStateProps: () => {
          return stateProps;
        },
        getActionsApi: () => {
          return {
            updateFieldValue: (propName, theValue) => {
              setValue(theValue);
            },
            triggerFieldChange: () => { /* nothing */}
          };
        },
        getValidationApi: () => {
          return {
            validate: () => { /* nothing */}
          };
        }
      };
    }
  };

  return (
    <>
      <PegaDxilMyBoolean {...props} {...args} />
    </>
  );
};

BasePegaDxilMyBoolean.args = {
  label: configProps.label,
  helperText: configProps.helperText,
  caption: configProps.caption,
  testId: configProps.testId,
  readOnly: configProps.readOnly,
  disabled: configProps.disabled,
  required: configProps.required,
  status: configProps.status,
  hideLabel: configProps.hideLabel,
  displayMode: configProps.displayMode,
  validatemessage: configProps.validatemessage
};
