// @pega/cosmos-react-rte/lib/components/RichTextEditor/RichTextEditor.styles.js
import styled, { css } from 'styled-components';
import { defaultThemeProp, StyledLink } from '@pega/cosmos-react-core';
import { listStyles } from '@pega/cosmos-react-core/lib/components/HTML/HTML';
export const StyledEditorContainer = styled.div(({ containerHeight, theme }) => {
    return css `
    padding: ${theme.base.spacing};

    ${typeof containerHeight === 'string' &&
        css `
      height: ${containerHeight};
    `}

    ${typeof containerHeight === 'object' &&
        css `
      ${containerHeight.min &&
            css `
        min-height: ${containerHeight.min};
      `}

      ${containerHeight.max &&
            css `
        max-height: ${containerHeight.max};
      `}
    `}

      &:hover {
      cursor: text;
    }

    ${listStyles}
  `;
});
StyledEditorContainer.defaultProps = defaultThemeProp;
const StyledRichTextEditor = styled.div(({ theme, focused }) => {
    return css `
    background-color: ${theme.base.palette['primary-background']};
    width: 100%;
    position: relative;
    border-color: ${theme.components['form-control']['border-color']};
    box-shadow: none;

    /* Override default styling of slate placeholder. */
    span[data-slate-placeholder='true'] {
      position: absolute;
      inset-block-start: 0;
      opacity: ${theme.base.transparency['transparent-3']} !important;
    }

    ${focused &&
        css `
      &:not([disabled]) {
        border-color: ${theme.components['form-control'][':focus']['border-color']};
        box-shadow: ${theme.components['form-control'][':focus']['box-shadow']};
      }
    `}
  `;
});
StyledRichTextEditor.defaultProps = defaultThemeProp;
export const StyledRichTextViewer = styled.div `
  width: 100%;
  position: relative;
  overflow-wrap: break-word;

  ${StyledLink} {
    &:enabled:focus,
    &:not([disabled]):focus {
      box-shadow: none;
      text-decoration: underline;
    }
  }
`;
export default StyledRichTextEditor;
