// @pega/cosmos-react-core/lib/components/DateTime/DateTime.test-ids.js
import { createTestIds } from '../../utils';
import { elements } from '../FormField/FormField.test-ids';
export const getDurationInputTestIds = createTestIds('duration-input', [
    'control-hour',
    'control-minute',
    'control-second',
    'control-number',
    'control-unit',
    ...elements
]);
export const getDateInputTestIds = createTestIds('date-input', [
    'control-day',
    'control-month',
    'control-year',
    'open-close-picker',
    'clear-current-date',
    'set-current-date',
    ...elements
]);
export const getDateRangeInputTestIds = createTestIds('date-range-input', [
    'from-date',
    'to-date',
    ...elements
]);
export const getDateTimeInputTestIds = createTestIds('date-time-input', [
    'control-day',
    'control-month',
    'control-year',
    'control-hour',
    'control-minute',
    'control-second',
    'control-period',
    'open-close-picker',
    'clear-current-date-and-time',
    'set-current-date-and-time',
    ...elements
]);
export const getDayOfWeekInputTestIds = createTestIds('day-of-week-input', [
    'control',
    ...elements
]);
export const getMonthInputTestIds = createTestIds('month-input', [
    'control-month',
    'control-year',
    ...elements
]);
export const getQuarterInputTestIds = createTestIds('quarter-input', [
    'control-quarter',
    'control-year',
    ...elements
]);
export const getTimeInputTestIds = createTestIds('time-input', [
    'control-hour',
    'control-minute',
    'control-second',
    'control-period',
    'open-close-picker',
    ...elements
]);
export const getTimeRangeInputTestIds = createTestIds('time-range-input', [
    'from-time',
    'to-time',
    'from-date-time',
    'to-date-time',
    ...elements
]);
export const getWeekInputTestIds = createTestIds('week-input', [
    'control-week',
    'control-year',
    'open-close-picker',
    ...elements
]);
export const getDatePickerTestIds = createTestIds('date-picker', [
    'previous-month',
    'month',
    'next-month',
    'calendar',
    'time'
]);
